Fix #9974 and part of #10069: Migration jobs and domain methods to populate UserAuthModel.

sajalasati opened the pull request:
## Overview

1. This PR fixes #9974 and part of #10069.
2. This PR does the following:

A) Added additional attributes in `UserAuthModel`. Now the UserAuthModel looks like:

| Attribute                                              | Full user | Profile user |
|--------------------------------------------------------|-----------|--------------|
| user_id (instances of this class are keyed by user_id) | Yes       | Yes          |
| gae_id                                                 | Yes       | None         |
| PIN (optional for both)                                | Yes       | Yes          |
| parent_user_id                                         | None      | Yes          |

None of the above attributes are **required** (except user_id which will be the `id`attribute). Whether they should be None or not will be ensured by external validation.

B) Added domain methods to populate UserAuthModel for new users.

C) Wrote migration jobs needed to populate the **"UserAuthModel"** (for the old users).

D) Put TODO(#10178) in UserSettings domain and storage classes for deprecating `gae_id` attribute.

E) Fixed a bug in `user_services.create_new_user`:
**Problem**: Current implementation passes `gae_id` into `get_user_settings` function to check whether a user is already registered or not ([here](https://github.com/oppia/oppia/blob/develop/core/domain/user_services.py#L853)). It's incorrect, because get_user_settings function expects a **user_id** NOT **gae_id**. Current code can "possibly" create multiple users with same gae_id and different user_ids. I tested it via a [a test method](https://github.com/oppia/oppia/blob/develop/core/domain/user_services_test.py#L86) in user_services_test file - called create_new_user with same gae_id - it produced different users with same gae_id and different user_id.
Why it is not a problem **currently**: Because in the BaseHandler does an additional check by `gae_id` of the user using `get_user_settings_by_gae_id` function ([here](https://github.com/oppia/oppia/blob/develop/core/controllers/base.py#L189)). But all the tests working on create_new_method like [these](https://github.com/oppia/oppia/blob/develop/core/domain/user_services_test.py#L86) are basically under false assumption.
**Solution**: Use user_services.get_user_settings_by_gae_id function inside create_new_user function. This fixes all tests using this function and also makes the function more reliable for production.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

Commit:
added code for class

Commit:
Merge branch 'develop' into adding-UserAuthModel

Commit:
complete class code

Commit:
removed incomplete test [skip ci]

Commit:
fix lint errors [skip ci]

Commit:
[skip ci] commit to skip ci

Commit:
resolve merge conflict

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
[skip ci] test added

Commit:
[skip ci] test added - fix lint error

Commit:
[skip ci] fixed test error

Commit:
[skip ci] all tests done

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
fix lint errors

Commit:
fix lint errors

Commit:
remove GAE constant from constants.ts file

Commit:
fix more lint errors

Commit:
addressed comments

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
added more tests

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
added validator class

Commit:
added test

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
fixed error

Commit:
fixed linting

Commit:
fixed linting

Commit:
nit changes

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
nit changes

Commit:
add constant

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
added changes and tests

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
rename test and add constant

Commit:
removed old constant

Commit:
fix takeout service test

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
addressed comments

Commit:
name changes

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
lint error fix

Commit:
addressed comments

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
added migration job and tests

Commit:
fix linting

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,41 @@ def reduce(key, version_and_exp_ids):
+    def map(item):
+        """Implements the map function for this job."""
+        user_id = item.id
+        gae_id = [item.gae_id]
```
``gae_ids`` since given that this is a list, it's plural in nature.

vinitamurthi replied:
But I think we shouldn't be using multiple gae_ids because we do not have the ability to map multiple email ids to a single account right now

sajalasati replied:
gae_id has been updated to single value attribute in PR #10004 and hence changed here also now.

BenHenning replied:
Thanks for updating.

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,41 @@ def reduce(key, version_and_exp_ids):
+                gae_id=gae_id
+            ).put()
+        elif user_auth_model.id != user_id:
+            # Because <user_id, gae_id> tuple coming from UserSettingsModel
```
Doesn't this case mean that a GAE ID is being shared between multiple users? I think we should trigger a failure in this case since it should never happen.

vinitamurthi replied:
A GAE ID can be shared between one user an its profiles. Since they are stored as user models they can be shared, however in our migration we should verify that two full accounts do not share the same gae_id, only profiles and their corresponding full account can share it.

sajalasati replied:
The code and the tests for it have been modified now. PTAL!

BenHenning replied:
That association doesn't actually make sense to me. See my other comment thread--I think this is going to cause problems when we actually start using the auth model for login.

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,41 @@ def reduce(key, version_and_exp_ids):
                     edited_exploration_ids))
 
 
+class UserAuthModelOneOffJob(jobs.BaseMapReduceOneOffJobManager):
```
What's the plan for integrating with the new auth model? Do we need a feconf flag for conditioning downstream code to depend on UserAuthModel vs. the user settings model for before this migration can be run in production?

vinitamurthi replied:
I think the plan is to use a flag

sajalasati replied:
@BenHenning Yes, the flag would be introduced for that conditioning. I wasn't sure where I should create it - now or in a later PR when we would actually use UserAuthModel or its methods.
@BenHenning or @vinitamurthi Pls confirm that just creating a constant in constant.ts file is enough for now? (As we don't have to condition anything based on it right now)

vinitamurthi replied:
I think you can add the flag when you actually use it. This PR is only creating the model so its not necessary to add it here.

sajalasati replied:
Thanks for clarifying that!

BenHenning replied:
Thanks for clarifying the plan & clarifying in the PR description.

BenHenning commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,63 @@ def test_user_contributions_get_created_after_running_the_job(self):
+        self.assertEqual(user_auth_model.gae_id[0], new_gae_id)
+        self.assertEqual(user_auth_model.id, user_settings_model.id)
+
+    def test_existing_gae_id_mapped_to_different_user_id_is_updated(self):
```
Reiterating above--I'd expect this case to fail, not remap.

sajalasati replied:
Fixed it in the code. Also the test `test_one_off_job_gae_id_registered_for_another_user_raises_error` now raises an error for this condition. PTAL!

BenHenning commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,63 @@ def test_user_contributions_get_created_after_running_the_job(self):
+        user_settings_model = user_models.UserSettingsModel.get_by_id(
+            self.user_a_id)
+        new_gae_id = user_settings_model.gae_id
+        self.assertIsNone(user_models.UserAuthModel.get_by_auth_id(
```
This test is actually verifying two behaviors:
1. That prior to running the migration job, no auth model exists for the given user ID
2. That after running the migration job, an auth model now exists for the user ID

sajalasati replied:
Separated the 2 tests now.

BenHenning commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,63 @@ def test_user_contributions_get_created_after_running_the_job(self):
             ['exp_id'])
 
 
+class UserAuthModelOneOffJobTests(test_utils.GenericTestBase):
```
What about a test case for running the migration job multiple times?

sajalasati replied:
Please see the updated tests now. As mentioned in PR description, this migration job is only supposed to be run once. I have written tests `test_one_off_job_gae_id_registered_for_another_user_raises_error` and `test_one_off_job_profile_registered_with_another_gae_id_raises_error` - there I have run migration jobs twice to check if exception is raised correctly or not. PTAL and tell if that fulfils the need.

BenHenning replied:
Note that map reduce jobs can fail midway due to unexpected reasons. You need to make sure your job is resilient to this--it should be able to be restarted and finish the migration successfully even if some data was already migrated. Migration jobs should also be idempotent--if they iterate over migrated data, that data shouldn't change no matter how many subsequent runs occur.

BenHenning reviewed (changes requested):
Thanks @sajalasati--left some thoughts & questions.

ankita240796 commented on core/domain/prod_validation_jobs_one_off_test.py:
```diff
@@ -13062,6 +13062,68 @@ def test_missing_edited_exploration_model_failure(self):
+            update_datastore_types_for_mock_datetime()
+            run_job_and_check_output(self, expected_output)
+
+    def test_audit_missing_user_settings_model_fails(self):
```
audit_with_missing ..

sajalasati replied:
Done in the original PR #10004.

ankita240796 reviewed (approved):
LGTM for prod jobs, just one minor nit. Thanks @sajalasati

seanlip commented:
@sajalasati For this PR, please assign me once @BenHenning and @vinitamurthi have approved. Thanks!

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

Commit:
removed repeated from gae_id

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
Merge remote-tracking branch 'upstream/develop' into adding-UserAuthModel

Commit:
Merge branch 'adding-UserAuthModel' into migration-userauthmodel

Commit:
Merge branch 'migration-userauthmodel' of github.com:sajalasati/oppia into migration-userauthmodel

Commit:
removed unwanted attributes

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
fix merge conflict

Commit:
documentation changes

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
revert time stamp in oppia_takeout_data.zip file

Commit:
merge conflict fix

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

Commit:
corrected code

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
Merge branch 'change-user-settings-model' into migration-userauthmodel

sajalasati commented:
**Reviewers please look only upto the commit** `50051be`. 
The commit `55f3b45` was done for merging, due to dependency on PR #10105.
Please see the updated [PR description](https://github.com/oppia/oppia/pull/10123#issue-459596113).

sajalasati commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,130 @@ def test_user_contributions_get_created_after_running_the_job(self):
+        self._run_one_off_job()
+        self.user_b_model.gae_id = self.user_a_model.gae_id
+        self.user_b_model.put()
+        with self.assertRaisesRegexp(
```
@ everyone!
Need help in the failing backend tests:
- `test_one_off_job_gae_id_registered_for_another_user_raises_error` and 
- `test_one_off_job_profile_registered_with_another_gae_id_raises_error`,

I am unable to understand [error message](https://github.com/oppia/oppia/pull/10123/checks#step:7:551) in the failing test. I have a `assertRaisesRegexp` condition to check whether the exception was raised or not. It seems that the exception isn't even being raised due to the error if I remove the `assertRaisesRegexp` in the code. Locally also I get a similar error like [this](https://github.com/oppia/oppia/pull/10123/checks#step:7:526) - `WARNING:root:Slice 1558342726402testbed-request-id-hash-0 0 failed for the 1 of up to 11 attempts (1 of 31 taskqueue execution attempts). Will retry now.`
I am unable to find the bug. If anyone can help, it would be great. Thanks!

BenHenning replied:
Tasks are run via a task queue. If a single task fails, it will print an error, not bubble the exception up to the test. See how other tests analyze the output of a test to verify if it failed correctly.

Note that you should probably be yielding an error in your job, not raising an exception. A raised exception means the task failed unexpectedly, so the task queue will try to restart the task. What you want to do is output an expected failure--the task succeeded with a detected failure case.

sajalasati replied:
Thanks @BenHenning for the help! I have used the class functionalities to get the error now. And the tests passed locally.

nithusha21 reviewed (approved):
LGTM for codeowner file

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -94,26 +94,38 @@ def entity_classes_to_map_over(cls):
+            all_user_profile_ids = item.associated_profile_user_ids
+            all_user_profile_ids.append(item.id)
+
+            user_auth_models = user_models.UserAuthModel.get_by_auth_id(
```
The documentation & impl of ``get_by_auth_id`` suggests that it returns a single model, not multiple. Does this code mean we need to change the contract of that method?

sajalasati replied:
I have updated the implementation of `get_by_auth_id` method in this PR - to use `fetch` instead of `get`, because profiles associated with the account will also be stored there and all `profile users` + `full user` have the same `gae_id`. `fetch` method in the implementation will fetch all the users stored with that `gae_id`.
(My initial thinking was that we won't be storing profiles there, `UserAuthModel` will just be used during gae login for full users, hence I had implemented that way).

**Update**: We will have only full user mapped with the gae_id. There won't be changes in the current implementation of `get_by_auth_id` method. These changes will be reverted.

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -94,26 +94,38 @@ def entity_classes_to_map_over(cls):
-                gae_id=gae_id
-            ).put()
+
+        # Assumption: full user_id + all its associated profile user_ids will
```
This doesn't quite make sense to me. How will GAE login work in this case? Isn't the expectation that you log into Oppia via a Google account through app engine, receive a GAE ID, then lookup the user to log into? How do we know which one to pick if there's a 1:many relationship from GAE ID to user ID?

Per the other comment thread, I was expecting that profiles would authenticate using a user ID + PIN, not a GAE ID. The idea there is that a profile logs in by providing a PIN and a valid user ID that's already been logged in via some other method (such as GAE ID).

It's also the case that there will be profiles associated with accounts that do not use GAE to auth (such as phone numbers).

sajalasati replied:
Ok, now I get what you meant by storing PIN in `UserAuthModel`. Please confirm if I am understanding it correctly: 

In the UserAuthModel we have:
- full user: user_id, gae_id, pin(maybe None)
- profile user: user_id, full_user_id, pin(maybe None)

Post-authentication flow in the backend:
- a full user sends gae_id in the request - we lookup by gae_id in the UserAuthModel
- a profile user sends its own user_id + pin + full_user_id in the request - correct full user id assures us that the user has been authenticated(right? This is the part I am not sure of) i.e existence of <full_user_id, user_id, pin> entry in the UserAuthModel.

BenHenning replied:
That's close to what I was thinking, but I just want to clarify a couple things:
1. Are you planning to keep 'gae_id' and the parent 'full_user_id' distinct?
2. For authenticating profiles, everything you mentioned matches what I was thinking except we need to verify that the parent full user ID is correctly authenticated. My suggestion is to not do any additional storage work to accommodate this--we need to figure this out during the authentication project (@anandwana001 FYI).

sajalasati replied:
1) Yes gae_id different from full_user_id. Following table shows the intent:

| Attribute                                              | Full user | Profile user |
|--------------------------------------------------------|-----------|--------------|
| user_id (instances of this class are keyed by user_id) | Yes       | Yes          |
| gae_id                                                 | Yes       | None         |
| PIN (optional for both)                                | Yes       | Yes          |
| parent_user_id                                         | None      | Yes          |

None of the above attributes are **required** (except user_id which will be the `id`attribute). Whether they should be None or not will be ensured by external validation.

2) Okay, as I will keep it to this much only.

BenHenning reviewed (changes requested):
Thanks @sajalasati. I had some high-level feedback and didn't dig into the tests yet.

sajalasati commented:
@BenHenning Now since UserSettingsModel does not have a associated profile user ids attribute, it's totally impossible to have a profile registered in our database - fully implementing profiles had other dependencies too (like auth etc), but even if they were ready it's impossible to support profiles before UserAuthModel is in use now right? Hence I can now safely assume that we will only have full users to be migrated to our new UserAuthModel.
(The original plan about the process as mentioned in PR description "Overview" is still valid).

Commit:
UserAuthModel changes and One off migration job

Commit:
added tests

Commit:
tests completed

Commit:
fix merge conflict

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

Commit:
fix linting

Commit:
fix leftover linting issues

sajalasati commented:
@vinitamurthi @BenHenning @vojtechjelinek PTAL now! I have modified the migration job now. Also updated the PR description.

vinitamurthi commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,43 @@ def reduce(key, version_and_exp_ids):
+            yield ('GAE_ID %s already registered with user_id %s.' % (
+                item.gae_id, user_auth_model.id), 1
+                  )
+        else:
```
We should add to the UserAuthModel only if the instance doesn't already exist (i.e. if id as well as gae_id match)

sajalasati replied:
Done.

vinitamurthi commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,43 @@ def reduce(key, version_and_exp_ids):
+            ).put()
+
+    @staticmethod
+    def reduce(error_message, _):
```
It may be better to mention the user ids of the error records

sajalasati replied:
Done.

vinitamurthi commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,43 @@ def reduce(key, version_and_exp_ids):
+            feconf.AUTH_METHOD_GAE, item.gae_id)
+        if (user_auth_model is not None) and (user_auth_model.id != item.id):
+            yield ('GAE_ID %s already registered with user_id %s.' % (
+                item.gae_id, user_auth_model.id), 1
```
Maybe return the user id / gae id combination as values instead of the '1'

sajalasati replied:
Done. PTAL now.

BenHenning commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,79 @@ def test_user_contributions_get_created_after_running_the_job(self):
+                self.USER_A_GAE_ID, self.user_a_model.id))
+        self.assertEqual(expected_error, output[0])
+
+    def test_one_off_job_run_multiple_times_raises_no_error(self):
```
Shouldn't you be checking that there are no expected errors from the output of the jobs? It seems like this test will always pass as-is.

sajalasati replied:
Did that check now. PTAL!

BenHenning replied:
Looks good--thanks!

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,43 @@ def reduce(key, version_and_exp_ids):
                     edited_exploration_ids))
 
 
+class UserAuthModelOneOffJob(jobs.BaseMapReduceOneOffJobManager):
```
Perhaps: ``PopulateUserAuthModelOneOffJob`` to be more specific.

sajalasati replied:
Renamed it now.

BenHenning reviewed (approved):
Thanks @sajalasati! Overall, the PR LGTM. Please address my last two comments, but otherwise I defer to @vinitamurthi to verify or I can if it's sent back to me.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -2228,6 +2226,9 @@ class UserAuthModel(base_models.BaseModel):
     # A code associated with profiles on android to provide a PIN based
     # authentication within the account.
     pin = ndb.StringProperty(default=None)
+    # User id of the full user which the profile user is associated with.
```
Perhaps explain in this comment why this info needs to be in this model (at first glance it has nothing to do with auth).

sajalasati replied:
Added more info now for the explaination.

seanlip replied:
See earlier comment for how to rewrite this.

Also -- this comment explains what the field is, but still doesn't explain why this is relevant for auth. Could you explain further?

seanlip reviewed (changes requested):
One question: when a new (full) user is created, will their UserAuthModel automatically be populated? Until that happens, you shouldn't merge this, because this job sounds like it will populate all UserAuthModels but then subsequent users who sign up after that job is run won't have those models defined.

sajalasati commented:
> One question: when a new (full) user is created, will their UserAuthModel automatically be populated? Until that happens, you shouldn't merge this, because this job sounds like it will populate all UserAuthModels but then subsequent users who sign up after that job is run won't have those models defined.

@seanlip This PR just defines the migration job. We actually use the UserAuthModel's methods and populate it in the PR #10176 (draft currently) - I am using a flag there, which prevents usage of UserAuthModel till it's enabled. Before the flag is set true, we will run this migration job and only after the setting of flag we use UserAuthModel. Both old and the new users will use UserAuthModel only to store their auth details afterwards.

seanlip commented:
OK, thanks for the context. It sounds like this PR should only go in after that stuff is in, then. 

To give some additional background: what usually happens in these situations is that the PR that flips the flag to True is the same PR that introduces the job. You would need to file a request to test the job following the instructions here: https://github.com/oppia/oppia/wiki/Running-jobs-in-production -- we'll then do a test run and if it passes we'll merge the PR, but merging the PR is conditional on a successful migration in production and subsequent verification that new signups are handled correctly from that point forward. This is because we will then follow the exact same set of instructions when doing the actual release in prod, and the merge of this PR marks the point when it is actually safe to transition. 

Also I would advise staging this more carefully. It would probably be better to do something like the following:

- Merge a PR such that UserAuthModels are automatically created for newly-signed-up users. At this point, some users will have UserAuthModels and some don't, so don't make use of the UserAuthModels yet.
- Merge this PR prior to a release, and fill out the form to run a job during the release to generate the missing UserAuthModels for existing users (it's fine to do this for all users too, I guess). None of your code should make use of UserAuthModels yet, though.
- The release cut should happen at this point. The code that is included in this release cut will, by virtue of the first point above, ensure that all newly-signed-up users have UserAuthModels. When the job is run in production, all users will have UserAuthModels.
- After the release cut is made, subsequent PRs can assume that all users have UserAuthModels and then you can start using them (and deprecating gae_id in UserSettingsModel etc.).

Note that you might not need a flag in that case. In fact I am not really sure what your flag in #10176 means -- usually we use flags to hide already-implemented behaviour until it's ready, but this is usually related to a feature rather than the details of the model implementation.

One tricky thing is that the release cut is this weekend. I think you might need to get it in by then, though. Otherwise if you make subsequent PRs that assume the existence of UserAuthModels for all users, things are going to break as soon as we make a deployment.

sajalasati commented:
@seanlip 
Regarding usage of flag: I think usage of UserAuthModel and its method also sounds like a feature to me in the sense that we don't just want to populate it, but it also affects our flow of creating a new user, how we fetch information on the basis of gae_id. I think we do need to introduce flags for that purpose. PR #10176 is almost ready now, so you may take a look at that. --All the methods to populate UserAuthModel or update/fetch data from it are behind that flag. Once the flag is enabled, only the new implementation (i.e via UserAuthModel) will be used, hence migrations should happen before we flip the flag.

seanlip commented:
Hi @sajalasati -- you haven't really addressed my comments, but just reiterated why you did what you did. The plan I proposed does not require taking the server offline (unless I'm misunderstanding something?), which seems better than the suggested migration with the flag, which does entail that. Could you comment on this?

The key here is that gae_id never changes after it is set, so once you've populated it in both systems then you can switch its usage from one system to the other without any issues.

vinitamurthi commented:
Hey @sajalasati , discussed this with @seanlip offline. Here's the general plan we can take:
1) In the future domain/controller level PRs, we will keep all cases where we read from the UserAuthModel behind a flag, creating UserAuthModels need not be behind a flag. This way new users would have their user data already populated in UserAuthModel.
2) This PR (i.e. the migration job) goes in only after we verify that new UserAuthModels are being generated correctly. After this point, the migration can happen.
3) When the feature is ready completely, we can flip the flag (after the migration is done)

sajalasati commented:
@vinitamurthi  Okay, thanks for clarifying that. I will make the changes accordingly.

Commit:
domain method to populate UserAuthModel

Commit:
modify migration job

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

Commit:
added todo

Commit:
fix linting

sajalasati commented:
@BenHenning @seanlip @vojtechjelinek @vinitamurthi  PTAL now! I have included the domain methods also in this PR now to include it in the release cut due tomorrow. Updated the PR description also to include all the tasks I am doing here. The plan is to follow the sequence of steps as @vinitamurthi mentioned in this [comment](https://github.com/oppia/oppia/pull/10123#issuecomment-670439324).

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,55 @@ def reduce(key, version_and_exp_ids):
+    """One-off job for creating and populating UserAuthModel for
+    all registered users.
+    """
+
```
Lets speed this job a bit by adding more shards.

```python
def enqueue(cls, job_id, additional_job_params=None):
    super(PopulateUserAuthModelOneOffJob, cls).enqueue(
		job_id, shard_count=64)
```

sajalasati replied:
Added this code. Is it fine now?

vojtechjelinek replied:
Yep

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,55 @@ def reduce(key, version_and_exp_ids):
+        # Assumptions:
+        # 1) We are only having full users in the database, and no
+        #    profile users.
+        # 2) For the same user_id, gae_id cannot be updated.
```
What does this mean?

sajalasati replied:
I thought of leaving these comments given our current situation (I wasn't sure whether to put them or not).
A bit of more explanation:
1) If we are having profiles also, for them we need to keep the `gae_id` field as empty. Currently we don't have a way of creating profiles, and this migration job would hence not have to deal with them (assuming this PR goes into release).
2) This migration works correctly only if gae_id cannot be updated in any way for a user. There isn't a way currently to update gae_id. 
I wrote these 2 points to just hint at the things which might be there in future, then migration job (if used then) would need modifications. Although they are not necessary to mention, I wanted to get others' opinion whether to write them or not.
wdyt @vojtechjelinek ?

vojtechjelinek replied:
Okay, I understood 1, but I get 2 now, could you please make it more understandable in the actual code comment?

sajalasati replied:
Sure I will write it properly.
Edit: Done now. PTAL!

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,55 @@ def reduce(key, version_and_exp_ids):
+                yield (
+                    'Error, gae_id already existing',
+                    {'user_id': item.id, 'gae_id': item.gae_id}
+                    )
```
```suggestion
                yield (
                    'Error, gae_id already existing',
                    {'user_id': item.id, 'gae_id': item.gae_id}
                )
```

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,55 @@ def reduce(key, version_and_exp_ids):
+                    )
+            else:
+                if (item.deleted is True) and (
+                        user_auth_model.deleted is not True):
```
```suggestion
                if (item.deleted is True and 
                	user_auth_model.deleted is not True):
```

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,145 @@ def test_user_contributions_get_created_after_running_the_job(self):
+        expected_output = [
+            {'user_id': self.USER_A_ID, 'gae_id': self.USER_A_GAE_ID},
+            {'user_id': self.USER_B_ID, 'gae_id': self.USER_B_GAE_ID},
+            ]
```
```suggestion
        expected_output = [
            {'user_id': self.USER_A_ID, 'gae_id': self.USER_A_GAE_ID},
            {'user_id': self.USER_B_ID, 'gae_id': self.USER_B_GAE_ID},
        ]
```

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,145 @@ def test_user_contributions_get_created_after_running_the_job(self):
+            {'user_id': self.USER_B_ID, 'gae_id': self.USER_B_GAE_ID},
+            ]
+
+        # To ensure both lists have same elements.
+        for user in expected_output:
+            self.assertIn(user, output['Created'])
+        for user in output['Created']:
+            self.assertIn(user, expected_output)
+        self.assertEqual(['Created'], output.keys()) # pylint: disable=dict-keys-not-iterating
```
```suggestion
        self.assertItemsEqual(output['Created'], expected_output)
        self.assertItemsEqual(['Created'], output.keys())
```

vojtechjelinek replied:
Maybe even just 
```suggestion
        self.assertItemsEqual(output, expected_output)
```
might work if you modify the `expected_output`

sajalasati replied:
Used a different way, but removed the extra code.

vojtechjelinek commented on core/domain/user_services.py:
```diff
@@ -850,19 +921,71 @@ def create_new_user(gae_id, email):
     user_settings = UserSettings(
         user_id, gae_id, email, feconf.ROLE_ID_EXPLORATION_EDITOR,
         preferred_language_codes=[constants.DEFAULT_LANGUAGE_CODE])
+    user_auth = UserAuth(user_id, gae_id)
+    _save_user_auth_details(user_auth)
```
_save_user_auth_details(UserAuth(user_id, gae_id))

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_services.py:
```diff
@@ -1118,6 +1241,11 @@ def mark_user_for_deletion(user_id):
     _save_user_settings(user_settings)
+    user_auth = _get_user_auth_from_model(
+        user_models.UserAuthModel.get_by_id(user_id))
+    if user_auth is not None:
+        user_auth.deleted = True
+        _save_user_auth_details(user_auth)
```
Maybe leave a TODO for later when we will always have both `UserSettingsModel` and `UserAuthModel` to remove the `None` check?

sajalasati replied:
Done.

vojtechjelinek reviewed (changes requested):
Thanks! Few comments.

Commit:
fix prod validation tests

sajalasati commented:
@ everyone! 
The backend tests are currently failing due to:
- failing of `core.domain.prod_validation_jobs_one_off_test.UserAuthModelValidatorTests` - this I have fixed (yet to push the change).
- Some IO error shown as below (I ran it locally also and it ran perfectly fine):
```
======================================================================
ERROR: test_ts_expect_error_error (scripts.linters.js_ts_linter_test.JsTsLintTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/oppia/oppia/scripts/linters/js_ts_linter_test.py", line 644, in test_ts_expect_error_error
    True).perform_all_lint_checks()
  File "/home/runner/work/oppia/oppia/scripts/linters/js_ts_linter.py", line 1161, in perform_all_lint_checks
    self.parsed_js_and_ts_files = self._validate_and_parse_js_and_ts_files()
  File "/home/runner/work/oppia/oppia/scripts/linters/js_ts_linter.py", line 192, in _validate_and_parse_js_and_ts_files
    file_content = self.file_cache.read(compiled_js_filepath)
  File "/home/runner/work/oppia/oppia/scripts/linters/pre_commit_linter.py", line 167, in read
    return self._get_data(filepath, mode)[0]
  File "/home/runner/work/oppia/oppia/scripts/linters/pre_commit_linter.py", line 198, in _get_data
    with python_utils.open_file(filepath, mode, newline='') as f:
  File "/home/runner/work/oppia/oppia/python_utils.py", line 103, in open_file
    raise IOError('Unable to open file: %s' % filename)
IOError: Unable to open file: /home/runner/work/oppia/oppia/tmpcompiledjs/scripts/linters/test_files/invalid_ts_expect_error.js
```

Commit:
address comments

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

sajalasati commented:
@vojtechjelinek Addressed your comments now. PTAL!

seanlip commented on core/domain/prod_validation_jobs_one_off_test.py:
```diff
@@ -13091,19 +13091,18 @@ def setUp(self):
-            pin=self.USER_PIN
-        ).put()
+
+        # Note: There will a total of 2 UserSettingsModel (hence 2
```
will --> will be

UserSettingsModel --> UserSettingsModels
UserAuthModel --> UserAuthModels

sajalasati replied:
Done.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,64 @@ def reduce(key, version_and_exp_ids):
+    def map(item):
+        """Implements the map function for this job."""
+
+        # Assumptions:
```
Probably should add a TODO + issue to remove this job once it has been run.

In the migration instructions you file for this job, also add a post-check to confirm that the number of models for the various different types are the same.

sajalasati replied:
Added a TODO.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,64 @@ def reduce(key, version_and_exp_ids):
+        #    the gae_id attribute is None for them.
+        # 2) This migration works correctly only if gae_id cannot be updated
+        #   in any way for a user. If a gae_id is mapped to some other user by
+        #   deleing it, an error will be raised while migration (see below).
```
deleing --> deleting
migration --> migrating

What does "it" refer to? Please be more specific -- I don't understand this last sentence. Also I don't think deleting stuff can cause the gae_id to be remapped -- it is auto-assigned by App Engine.

sajalasati replied:
@seanlip Sorry for the poor language.
In the last sentence I wanted to present this scenario - We have done a migration once. There is a current user A whose entry exists in the UserAuthModel. A deletes his account (he had a `gae_id` assigned to himself), and new user B signs up with the same google email id - then if there is a possibility that the same `gae_id` is mapped to the new user B also (coming from App engine because he has the same email id), then while migrating users I will raise an error saying the same `gae_id` is registered to some other user in the UserAuthModel.

But this is an incomplete thought/analysis.

My further thinking on it:
1) If a UserAuthModel entry exists for user A (which can be post migration only), then while deleting user A from UserSettingsModel, we will also mark the corresponding UserAuthModel entry for deletion (this is what the current code will enforce).
2) When a user is deleted, I am not sure what happens with the <gae_id, email id> mapping in the App engine itself. Is it deleted from their records too? Can it happen that a new user comes and sign up with the same gae_id? -- this was a question in my mind, but when I think clearly, it doesn't really affect me - because of point (1) - UserSettingsModel or UserAuthModel entries would be properly deleted when the user is deleted, so we would basically have no record of that gae_id.

Outcome of all this discussion:
- Finally the comment would look like: `This migration works correctly only if gae_id once mapped to a user cannot be updated in any way. Currently there is no way to do so, but the migration job would need modifications if it has to work in such a setting in future.`

seanlip replied:
Thanks for the explanation. To answer your question 2, it is the case that that mapping stays the same (it's not maintained within our app).

The final comment looks good. Some wordsmithing: `This migration assumes that, once a gae_id is assigned to a user, it cannot be updated. Currently, there is no way to do so, but the migration job would need modifications if it has to work in such a setting in future.` (Though hopefully we would not need to keep it around that long -- we should delete it as soon as it is run in order to avoid it being run erroneously.)

sajalasati replied:
Did the change now.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,64 @@ def reduce(key, version_and_exp_ids):
+        if user_auth_model is not None:
+            if user_auth_model.id != item.id:
+                yield (
+                    'Error, gae_id already existing',
```
ERROR - gae_id already exists

sajalasati replied:
Not required now. Please see the updated migration job.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,64 @@ def reduce(key, version_and_exp_ids):
+                    user_auth_model.deleted = item.deleted
+                    user_auth_model.put()
+                    yield (
+                        'Updated', {'user_id': item.id, 'gae_id': item.gae_id})
```
The key is not very specific. Please use something clearer, e.g. "ERROR - Deleted UserAuth model."

sajalasati replied:
Not required now. Please see the updated migration job.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,64 @@ def reduce(key, version_and_exp_ids):
+                gae_id=item.gae_id,
+                deleted=item.deleted
+            ).put()
+            yield ('Created', {'user_id': item.id, 'gae_id': item.gae_id})
```
'SUCCESS - Created UserAuth model'

The second arg should be a string (see other jobs for examples).

vojtechjelinek replied:
I think it can also be a tuple or some other object, I think I used tuples in the user ID migration.

seanlip replied:
Oh, sorry -- thanks for catching that @vojtechjelinek, I was thinking of reduce(). Yes this is fine but if you do that, convert it to a string in reduce(). (Having said that it is probably easier to just use a string here, given that your reduce() is trivial.)

sajalasati replied:
Returning just `1` now as we just want the number of users populated as per the other comment.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -81,6 +81,64 @@ def reduce(key, version_and_exp_ids):
+    def reduce(message_category, values):
+        users_list = [
+            ast.literal_eval(v) for v in values]
+        yield (message_category, users_list)
```
For the success cases, yield the length. Only yield the actual result for the error cases.

sajalasati replied:
Done. There is only one case now i.e for success.

seanlip commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,141 @@ def test_user_contributions_get_created_after_running_the_job(self):
+        }
+        self.assertEqual(output, expected_output)
+
+
```
Drop extra newline (separate adjacent methods in a class by a single newline)

sajalasati replied:
Done.

seanlip commented on core/domain/user_services.py:
```diff
@@ -47,6 +47,9 @@
     'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEwAAABMCAYAAADHl1ErAAAAAXNSR0IArs4c6QAADhtJREFUeAHtXHlwVdUZ/859jyxmIQESyCaglC0iAgkJIntrIpvKphSwY2ttxbFOp9R/cGGqdhykLaMVO2OtoyRSCEKNEpYKyBIVQ1iNkBhNMCtb8shiQpJ3b7/fTW7m5uUlecu9L4nTM5Pce8895zvf93vnnPud833fEdQLKXb5jsC6%2BuZERZbHKaSMYRbGKERxgpQQUkSIIigEbAmFavlfrUKiVhCVcFa%2BIJEvJOlCcNCAnNKMFQ0o58vEfPgmhS5Mn0ot8n2KIs8lIZJJUfy8almIJqbxhRDSIbJKe2s%2BXv ...
 
 
+# TODO(#10178): gae_id will be deprecated for UserSettings because
```
Frame TODOs as imperatives. I.e.:

"Deprecate gae_id for UserSettings once we have verified that UserAuthModels exists for every user."

(No need for the last sentence.)

sajalasati replied:
Done.

seanlip commented on core/domain/user_services.py:
```diff
@@ -295,6 +298,75 @@ def require_valid_username(cls, username):
+        user_id: str. The unique ID of the user.
+        gae_id: str. The ID of the user retrieved from GAE.
+        pin: str or None. The PIN of the user's profile for android.
+        parent_user_id: str or None. User id of the full user to which a
```
Change first sentence of description to: `For profile users, the user ID of the full user associated with that profile.`

Ditto elsewhere.

sajalasati replied:
Done. Please check if it explains properly why we need to store `parent_user_id` in UserAuthModel from authentication point of view.

seanlip commented on core/domain/user_services.py:
```diff
@@ -850,19 +922,70 @@ def create_new_user(gae_id, email):
     return user_settings
 
 
+def _save_user_auth_details(user_auth):
```
This is actually a better name. Could you name the domain object UserAuthDetails and the storage model UserAuthDetailsModel? Make sure to change variable/function names accordingly, too.

sajalasati replied:
Did the name change everywhere.

seanlip commented on core/domain/user_services.py:
```diff
@@ -1118,6 +1241,14 @@ def mark_user_for_deletion(user_id):
+    user_auth = _get_user_auth_from_model(
+        user_models.UserAuthModel.get_by_id(user_id))
+
+    # TODO(#10178): Remove the if condition below when UserAuthModel will
```
when UserAuthModel will be --> once UserAuthModel is

sajalasati replied:
Done.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -40,14 +40,16 @@
 
     # User id used to identify user by GAE. Is not required for now because we
     # need to perform migration to fill this for existing users.
+    # TODO(#10178): gae_id will be deprecated for UserSettingsModel because
```
See notes above re how to rewrite this.

sajalasati replied:
Done. Changed now.

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -2276,12 +2276,17 @@ def test_create_raises_error_when_many_id_collisions_occur(self):
     """Tests for UserAuthModel."""
 
-    NONEXISTENT_AUTH_TYPE_NAME = 'id_x'
+    NONEXISTENT_AUTH_METHOD_NAME = 'id_x'
```
Why call this "id_..."? Should be "auth_method_x" perhaps?

sajalasati replied:
Sorry I hadn't paid attention to it earlier. Changed now.

vinitamurthi replied:
This doesnt seem changed?

sajalasati replied:
Sorry fixed it now.

BenHenning commented on core/domain/prod_validation_jobs_one_off_test.py:
```diff
@@ -13091,19 +13091,18 @@ def setUp(self):
-            pin=self.USER_PIN
-        ).put()
+
+        # Note: There will a total of 2 UserSettingsModel (hence 2
```
Nit on wording: ``There will be a total of 2 UserSettingsModels``

sajalasati replied:
Done.

seanlip commented on core/domain/user_services.py:
```diff
@@ -850,19 +922,70 @@ def create_new_user(gae_id, email):
         Exception. If a user with the given gae_id already exists.
     """
-    user_settings = get_user_settings(gae_id, strict=False)
+    user_settings = get_user_settings_by_gae_id(gae_id, strict=False)
```
Great catch -- thanks @sajalasati! 

/cc @vojtechjelinek -- FYI this seems to have been missed during the migration. Can you confirm that this was an error?

vojtechjelinek replied:
Yes, this is an error. Looking at the code this shouldn't have caused any problems since in base.py we use `get_user_settings_by_gae_id` for checking if the user exists and only then call  `create_new_user`, and here since `gae_id` and `user_id` have different lengths it will always return `None`. Also, this and lines 926 to 929 can probably be replaced by `get_user_settings_by_gae_id(gae_id, strict=True)`.

sajalasati replied:
@vojtechjelinek why `strict=True`? --this flag is meant for raising an exception when we don't find UserSettings. Here we want to raise an exception inside `create_new_user` when a UserSettings does exist. I don't think we can change how the strict flag is used either because we don't know inside `get_user_settings_by_gae_id` when to raise an error when UserSettings is found. I would prefer the current implementation only.

vojtechjelinek replied:
Oh yeah, sorry my mistake, I just misread the code. The implementation is fine then.

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -95,27 +101,42 @@ def entity_classes_to_map_over(cls):
+        #    the gae_id attribute is None for them.
+        # 2) This migration works correctly only if gae_id cannot be updated
+        #   in any way for a user. If a gae_id is mapped to some other user by
+        #   deleing it, an error will be raised while migration (see below).
```
Wording: either ``during migration`` or ``while migrating``

sajalasati replied:
Changed the description now as @seanlip suggested in [this comment](https://github.com/oppia/oppia/pull/10123#discussion_r467410927)

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -95,27 +101,42 @@ def entity_classes_to_map_over(cls):
+                    {'user_id': item.id, 'gae_id': item.gae_id}
+                )
+            else:
+                if (item.deleted is True and
+                        user_auth_model.deleted is not True):
```
```suggestion
                if item.deleted and not user_auth_model.deleted:
```

sajalasati replied:
Not required now. Please see the updated migration job.

seanlip commented:
@sajalasati I took a look at the backend tests. Seems like the main failure now is a coverage issue?

BenHenning commented on core/domain/user_services.py:
```diff
@@ -295,6 +298,75 @@ def require_valid_username(cls, username):
+                self.pin
+            )
+
+        if (self.parent_user_id is not None and
```
We should also validate that:
- PIN and parent ID should always be defined together
- GAE ID and parent ID should never both be present

sajalasati replied:
Did the changes for:
> - GAE ID and parent ID should never both be present

But regarding:
> - PIN and parent ID should always be defined together.

I think this is not True. Consider a profile which has a parent id but not a PIN set yet or a full user on android with Parent ID as None but having a defined PIN - both are possible I believe.

BenHenning replied:
The latter is allowed, you’re correct. I don’t think the former is currently allowed, but I suppose that doesn’t mean we should prohibit it here.

BenHenning commented on core/domain/user_services.py:
```diff
@@ -850,19 +922,70 @@ def create_new_user(gae_id, email):
+
+    # If user auth entry with the given user_id does not exist, create a new
+    # one.
+    user_auth_model = user_models.UserAuthModel.get_by_id(user_auth.user_id)
```
Won't this raise an exception? I think ``get_by_id`` only returns ``None`` if you do a non-strict retrieval.

sajalasati replied:
Sorry I did not understand what you meant by non-strict retrieval.
This code does work without raising an exception. The tests have covered them. If the id `user_auth.user_id` does not exists in `UserAuthModel` then it will return None, else the model. This same implementation is there is `user_services._save_user_settings()` function also. Note `user_auth.user_id` itself will never be None here.

seanlip replied:
@BenHenning just a note, this is the model layer, which returns None if the NDB model instance is not found. See docs at https://cloud.google.com/appengine/docs/standard/python/ndb/modelclass#class_methods

BenHenning replied:
Ah thanks. I did conflate this with the base model get() function.

BenHenning commented on core/domain/user_services.py:
```diff
@@ -850,19 +922,70 @@ def create_new_user(gae_id, email):
+    """Transform user auth storage model to domain object.
+
+    Args:
+        user_auth_model: UserAuthModel or None. The model to be converted or
```
Why is None valid? Why not just expect that the caller ensures that the model is defined?

If 'None' is really what we want, we should define what it means for/what happens when 'None' is used.

sajalasati replied:
Actually I was trying to mimic the design of `user_services._transform_user_settings()`- [this](https://github.com/oppia/oppia/blob/develop/core/domain/user_services.py#L748). We can either do a check before sending the model or do it like the current implementation - after getting the domain object. Both the implementations styled are equally good in my opinion. It just depends where we want to put the if condition. If remove None inside this function seems like a better coding style, then I will go with that.

(Purpose was just that in the domain layer we will mostly deal with the UserAuth domain object, so instead of requiring an additional check before we convert it (whether is model is None or not), we will first convert storage object to domain object and then see. One analogous example for `UserSettings` is [this](https://github.com/oppia/oppia/blob/develop/core/domain/user_services.py#L512).)

seanlip replied:
I think this is similar to one of my other comments (possibly in a different PR) -- if the function conceptually is dealing with an actual model, then callers should pass in that model rather than "model or None". In this case I think callers should pass in an actual model.

I suggest updating the existing UserSettings functions you are copying in order to remove this bad pattern, so that others don't copy it in the future too. (This code was written a long time ago in a time where our codebase was less structured, so we probably ought to update it.)

Also, thanks for explaining your reasoning so clearly!

sajalasati replied:
@seanlip thanks for clarifying. I will do the changes.

sajalasati replied:
Did the changes in `user_services._transform_user_settings()` as well as `_save_user_auth_details` - removed None inside these functions and changed at code at related places.

BenHenning commented on core/domain/user_jobs_one_off.py:
```diff
@@ -95,27 +101,42 @@ def entity_classes_to_map_over(cls):
+                    {'user_id': item.id, 'gae_id': item.gae_id}
+                )
+            else:
+                if (item.deleted is True and
```
Please add tests for the deletion cases.

sajalasati replied:
Not required now. Please see the updated migration job.

BenHenning commented on core/domain/user_services_test.py:
```diff
@@ -1437,6 +1473,59 @@ def test_created_on_gets_updated_correctly(self):
         self.assertEqual(user_settings_model.created_on, time_of_creation)
 
 
+class UserAuthTests(test_utils.GenericTestBase):
```
Please make sure all validation error cases are tested.

sajalasati replied:
Done.

BenHenning reviewed (changes requested):
Thanks @sajalasati -- a lot has changed since my last review, so requesting some additional changes.

vinitamurthi commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -196,6 +196,141 @@ def test_user_contributions_get_created_after_running_the_job(self):
+        }
+        self.assertEqual(expected_output, output)
+
+    def test_one_off_job_existing_user_updated_successfully(self):
```
Do we need to add a test where an existing user does not get updated if they already have a gae_id/user_id combination in UserAuthModel ?

sajalasati replied:
Not required now. Please see the updated migration job.

vojtechjelinek reviewed (approved):
LGTM!

sajalasati commented:
@seanlip @vinitamurthi @vojtechjelinek @BenHenning PTAL at this! (Sorry for the length of this comment).
I paid closer attention to the migration job, and what sort of assumptions we can make. I think I can reduce it to this map function to this:
(I want to know if my logic seems reasonable, then I can go ahead and make tests acc to it -- I just wanted to avoid going back and forth on this. It looks weird that no error message is being produced, but I have a reason for that below)

```python
class PopulateUserAuthModelOneOffJob(jobs.BaseMapReduceOneOffJobManager):
    """One-off job for creating and populating UserAuthModel for
    all registered users.
    """

    @staticmethod
    def map(item):
        """Implements the map function for this job."""

        # Assumptions:
        # 1) We are only having full users in the database, and no
        #    profile users (For profiles we need to do a role check also and
        #    ensure the gae_id attribute is None for them).
        # 2) This migration assumes that, once a gae_id is assigned to a user, it
        #    cannot be updated. Currently, there is no way to do so, but the
        #    migration job would need modifications if it has to work in such a
        #    setting in future. 

        user_auth_model_by_gae_id = user_models.UserAuthModel.get_by_auth_id(
            feconf.AUTH_METHOD_GAE, item.gae_id)

        if user_auth_model_by_gae_id is None:
            user_models.UserAuthModel(
                id=item.id,
                gae_id=item.gae_id,
                deleted=item.deleted
            ).put()
            yield ('Created', {'user_id': item.id, 'gae_id': item.gae_id})
```

**Reason**: UserAuthModel exactly represents a subset of the UserSettingsModel (the fields of pin, parent_user id) at any point of time. This subset's entries are always up-to-date with the ones in UserSettingsModel.

**Explaination**:
- If a new user is created it's populated in both `UserSettingsModel` and `UserAuthModel`. `create_new_user` already ensures that no two users can be registered with the same `gae_id` (for old as well as new users). Hence for old users we are sure that no 2 users have the same gae_id. The same can be assured for new users too because they are populated in both the models.
- If a user is ever deleted from `UserSettingsModel`, we also delete the corresponding `UserAuthModel` entry if it exists. So it would never happen that an old deleted <user_id, gae_id> entry of UserSettingsModel is present in `UserAuthModel`.
- As mentioned in the code comment, we assume an update of gae_id cannot happen. I think its reasonable to assume, we won't allow it in future also.
- If any old user's entry is present in UserAuthModel, it can happen only via migration. Since `UserSettingsModel` will always maintain a state where 1 gae_id is for 1 user_id and vice versa, so it will not have any conflicts with the entries in UserAuthModel.

Hence running a migration job just becomes filling the new entries - basically updating a subset to become the whole set.

Previously in my code `Update` and `Error` cases were basically some impossible conditions. I am trying to make use of current code structure to simplify the job.

I just want to know about my logic -- I will make other changes also as mentioned in rest of the comments.

seanlip commented:
Yes, I was wondering why your job was so complicated -- this was more in line with the kind of thing I was envisioning. One other suggestion I have though is to just have the job simply regenerate the model for all users, regardless of whether the model already exists or not; that would take care of any issues with possibly-stale data.

vinitamurthi commented:
Yeah I think the job itself can be simple but I would like to see test cases for all of the things you mentioned above that verifies that the job doesn't affect those records.

vinitamurthi commented:
Or if you go with Sean's suggestion, that the job rewrites  old records.

sajalasati commented:
@seanlip @vinitamurthi thanks for a quick reply.
Yes the suggestion for re-writing does sound like a safer approach (although I cannot think of an case where we would have stale data). But then we will have to erase the entire UserAuthModel everytime right? Also I am not sure if writes are costlier than reads - @seanlip you would have already thought about it but still just mentioning.

@vinitamurthi sure I will write tests for them.

seanlip commented:
When you put() with the same ID, you overwrite the entity. (Check that manually, in order to make sure you're doing it correctly.) So no deletion is needed.

Writes are fine if we run this job just once. I'm not sure what you mean by "everytime", since if this has been designed properly you will need to run this job at most once.

sajalasati commented:
@seanlip 
Yes I will check that manually and with tests also. 
Yes, it is supposed to be run just once only. But if at all there is a case when we want to run it again, I meant that -- but fine, you can ignore that "everytime".

Commit:
modify migration

Commit:
addressed some comments

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

Commit:
address all comments

sajalasati commented:
@seanlip - addressed all your comments.
@BenHenning Follow up questions/remarks on some of your comments - [this](https://github.com/oppia/oppia/pull/10123#discussion_r467348243), [this](https://github.com/oppia/oppia/pull/10123#discussion_r467348153)
@vinitamurthi @vojtechjelinek PTAL!

vojtechjelinek commented:
Yep, I'm fine with the changes.

vinitamurthi reviewed (commented):
LGTM apart from the comments that are already covered by @seanlip and @BenHenning, please address their comments

vinitamurthi reviewed (approved):
I am giving an approval to unblock this in case the comments of @seanlip and @BenHenning  are addressed while I am not online but please make sure you address those comments

Commit:
added test

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

Commit:
changed name to user auth details

sajalasati commented:
Before the commit -- `added test` - `19b0f46`, all tests were passing. For this and further commits, following code change has been done:
- I have added one more test for migration job (in `user_jobs_one_off_test.PopulateUserAuthDetailsModelOneOffJobTests`) - to test whether `user_services.mark_user_for_deletion` works correctly for old users that have been migrated. With this @vinitamurthi I think I have covered all the test cases which I was talking about in [this comment](https://github.com/oppia/oppia/pull/10123#issuecomment-670853721) (you had asked to cover them in [your](https://github.com/oppia/oppia/pull/10123#issuecomment-670855917) comment), so you can see that change.

- I have done the name changes now everywhere - to use the term `UserAuthDetails` as @seanlip asked for it [here](https://github.com/oppia/oppia/pull/10123#discussion_r467347102).

seanlip commented on core/domain/user_services.py:
```diff
@@ -423,7 +504,10 @@ def get_users_settings(user_ids):
             ))
         else:
-            result.append(_transform_user_settings(model))
+            if model is not None:
```
Always keep the code being branched as small as possible. In this case, these 4 lines should have been: `result.append(_transform_user_settings(model) if model is not None else None)`.

sajalasati replied:
Done. Sure, I will use this style in future wherever applicable.

seanlip commented on core/domain/user_services.py:
```diff
@@ -525,12 +609,15 @@ def get_user_settings_by_gae_id(gae_id, strict=False):
-        raise Exception('User not found.')
-    return user_settings
+    user_settings_model = user_models.UserSettingsModel.get_by_gae_id(gae_id)
+    if user_settings_model:
```
if user_settings_model is not None

sajalasati replied:
Did the change now.

seanlip commented on core/domain/user_services.py:
```diff
@@ -525,12 +609,15 @@ def get_user_settings_by_gae_id(gae_id, strict=False):
+    if user_settings_model:
+        user_settings = _transform_user_settings(user_settings_model)
+        return user_settings
+    else:
```
```
elif strict:
    ...
else:
   return None
```

sajalasati replied:
Done.

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -1437,6 +1476,105 @@ def test_created_on_gets_updated_correctly(self):
+            self.user_auth_details.validate()
+
+    def test_validate_user_id(self):
+        self.user_auth_details.user_id = 'uid_' + 'a' * 31 + 'A'
```
Here and elsewhere: prefer string interpolation to concatenation. E.g. `'uid_%sA' % ('a' * 31)`

seanlip replied:
Note: this is mentioned in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide#python)

sajalasati replied:
Did the change here and elsewhere in this file too.

seanlip commented:
@sajalasati please fill out the [release form](https://github.com/oppia/oppia/wiki/Running-jobs-in-production) with instructions for the server admin if you want this to be run. Thanks!

seanlip commented:
(It is fine to do this now, assuming that you plan to fix these changes and get the PR in for the upcoming release.)

BenHenning reviewed (approved):
Thanks! LGTM

Commit:
address comments

Commit:
Merge remote-tracking branch 'upstream/develop' into migration-userauthmodel

Commit:
fix linting

seanlip reviewed (approved):
Thanks, LGTM. Please fill in the release form so that this can be tested and merged.

sajalasati commented:
@seanlip Made the changes you asked for. PTAL!

Regarding the form, is it necessary to fill it before the PR is merged? It's next on my list of tasks, but do you mind me filling it after some hours(need to take rest now) or should it be filled right now. Yes I do want to get this PR in the scheduled release. I will fill form as soon as I get back. Thanks!

seanlip commented:
I do need the form in order to run this on the server and validate that it
runs successfully before it can be merged, sorry.

More specifically, though, I just need the instructions doc, with
pre-checks and post-checks. If you can give me that, it will be sufficient
for now.

>

sajalasati commented:
@seanlip No problem, I will complete it now itself.

sajalasati commented:
@seanlip Link to instructions doc - [this](https://docs.google.com/document/d/1KYgY5AZRo5WRPb2kN6mR9foK3FdMojlh7H9FJgTsHbo/edit?usp=sharing).

seanlip commented:
Sg -- thanks! Will do a test run and, if it passes, get this merged.

seanlip commented:
The job succeeded! This PR is cleared for merge.

Please file the full release form ASAP; the release coordinators are currently planning the sequence of job runs for August.

sajalasati commented:
@seanlip @aks681 I have filled and submitted the release form few hours ago, forgot to mention it here. Please let me know if anything else is also needed from my side. Thanks!

seanlip commented:
Hi Sajal -- it's all good, your job is slated for the August release.
Nothing else is needed from your end. Thanks a lot!

sajalasati commented:
Thanks @seanlip !!
